### PR TITLE
[AND-5] Expand message theming

### PR DESCRIPTION
**Pull request by kanat**
### 🎯 Goal

Closes:
- https://linear.app/stream/issue/AND-5/expand-message-theming

#### General
- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] Assigned a person / code owner group (required)
- [ ] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [ ] PR targets the `develop` branch
- [ ] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_

### CODE REVIEW

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/messages/PollMessageContent.kt [hunk 1]
```diff
@@ -115,10 +119,9 @@ public fun PollMessageContent(
             true -> ChatTheme.ownMessageTheme.deletedBackgroundColor
             else -> ChatTheme.otherMessageTheme.deletedBackgroundColor
         }
-
         else -> when (ownsMessage) {
-            true -> ChatTheme.colors.linkBackground
-            else -> ChatTheme.otherMessageTheme.backgroundColor
+            true -> ChatTheme.ownMessageTheme.poll.backgroundColor
+            else -> ChatTheme.otherMessageTheme.poll.backgroundColor
```
Comment (kanat) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/components/messages/PollMessageContent.kt:124]: Replaced with `PollStyle.backgroundColor` to avoid unexpected color changes for polls when there is a change for `linkBackground` for instance.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/theme/messages/list/PollMessageStyle.kt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+         * Builds the default poll message style for the current user's messages.
+         *
+         * Returns a [PollMessageStyle] instance holding our default theming.
+         */
+        @Suppress("DEPRECATION_ERROR")
+        @Composable
+        public fun defaultStyle(
+            own: Boolean,
+            isInDarkMode: Boolean,
+            typography: StreamTypography = StreamTypography.defaultTypography(),
```
[the remark this answers is not in the document]
Comment (kanat) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/theme/messages/list/PollMessageStyle.kt:46]: it's for further extension of the Style, to avoid changing the public API :)

File: CHANGELOG.md [hunk 3]
```diff
@@ -51,6 +51,7 @@
 ### ✅ Added
 
 ### ⚠️ Changed
+- 🚨 Breaking change: The `MessagePositionHandler.handleMessagePosition` function now includes a new parameter, `isBeforeDateSeparator`, to handle message positioning more precisely within a group of messages. [#5466](https://github.com/GetStream/stream-chat-android/pull/5466)
```
Comment (kanat) [inline CHANGELOG.md:54]: There is a breaking change ⚠️

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/theme/MessageTheme.kt [hunk 4]
```diff
@@ -53,6 +83,7 @@ public data class MessageTheme(
         public fun defaultOwnTheme(
             isInDarkMode: Boolean = isSystemInDarkTheme(),
             typography: StreamTypography = StreamTypography.defaultTypography(),
+            shapes: StreamShapes = StreamShapes.defaultShapes(),
```
[the remark this answers is not in the document]
Comment (kanat) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/theme/MessageTheme.kt:86]: Mainly we just pass some common attributes like `colors`, `shapes`, etc., so any changes in root shapes would have effect on any underlying components.

If you need just replace `contentPadding`, then you can do something like this:
```kotlin
val ownTheme = MessageTheme.defaultOwnTheme().copy(
    contentPadding = ComponentPadding.all(8.dp),
)
```

File: stream-chat-android-ui-common/src/main/kotlin/io/getstream/chat/android/ui/common/feature/messages/list/MessagePositionHandler.kt [hunk 5]
```diff
@@ -56,31 +59,42 @@ public fun interface MessagePositionHandler {
...
+
+                    val isBottom = !isAfterDateSeparator &&
+                        (isBeforeDateSeparator || nextMessage.isInvalidMessageFromUser(curUser)) &&
+                        prevMessage.isValidMessageFromUser(curUser)
+
+                    if (isTop) add(MessagePosition.TOP)
+                    if (isMiddle) add(MessagePosition.MIDDLE)
+                    if (isBottom) add(MessagePosition.BOTTOM)
                     if (isEmpty()) add(MessagePosition.NONE)
                 }
```
Comment (kanat) [inline stream-chat-android-ui-common/src/main/kotlin/io/getstream/chat/android/ui/common/feature/messages/list/MessagePositionHandler.kt:88]: Fixed `MessagePosition` determination so that `NONE` represents the case when the message is a single isolated message, and addressed other corner cases.

File: stream-chat-android-ui-common/src/test/kotlin/io/getstream/chat/android/ui/common/feature/messages/list/MessagePositionHandlerTest.kt [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+import io.getstream.chat.android.models.Message
+import io.getstream.chat.android.models.MessageType
+import io.getstream.chat.android.models.User
+import io.getstream.chat.android.randomMessage
+import io.getstream.chat.android.ui.common.state.messages.list.MessagePosition
+import org.amshove.kluent.`should be equal to`
+import org.junit.jupiter.api.Test
+
+internal class MessagePositionHandlerTest {
```
Comment (kanat) [inline stream-chat-android-ui-common/src/test/kotlin/io/getstream/chat/android/ui/common/feature/messages/list/MessagePositionHandlerTest.kt:27]: `MessagePositionHandlerTest` covers various scenarios for `MessagePosition` determination.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/theme/MessageTheme.kt [hunk 7]
```diff
@@ -120,7 +174,43 @@ public data class MessageTheme(
...
+ * @param middle The shape which is used for the middle message in a group.
+ * @param bottom The shape which is used for the bottom message in a group.
+ * @param none The shape which is used for messages that are not in a group.
+ */
+public data class MessageBackgroundShapes(
+    val top: Shape,
+    val middle: Shape,
+    val bottom: Shape,
+    val none: Shape,
+)
```
Comment (kanat) [inline stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/theme/MessageTheme.kt:216]: We now have a distinct shape for each `MessagePosition`.

File: stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/theme/messages/list/PollMessageStyle.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (VelikovPetar) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/theme/messages/list/PollMessageStyle.kt:46]: Just for my understanding: Is it OK to release a version of the SDK with such unused configurable arguments? I am worried that a user might expect that those customisations are possible, but in the end changing them would do nothing? Or is it expected to also extend the style before releasing the next version?

[the remark this answers is not in the document]
Comment (kanat) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/theme/messages/list/PollMessageStyle.kt:46]: I believe it should be fine. They’re there to future-proof the API, so we don’t have to break it when customization requests for polls come in. Most customers tweak `colors`, `typography` and `shapes` in the root `ChatTheme` anyway, so it’s more about keeping things flexible just in case.

[the remark this answers is not in the document]
Comment (VelikovPetar) [reply stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/theme/messages/list/PollMessageStyle.kt:46]: Ok, thanks for the explanation!

**Review by JcMinarro — APPROVED** [review]
